Add upcoming/missed/due schedules indicator on budget table (based on category schedule templates)

joel-jeremy opened the pull request:
<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes. Try running yarn generate:release-notes *before* pushing your PR for an interactive experience. -->
Show upconming, due, or missed schedules to help users with budget decisions

![image](https://github.com/user-attachments/assets/45e4ca6d-6522-4c6e-a36c-3dcd253e0526)
![image](https://github.com/user-attachments/assets/5b4f905a-5c49-414c-b764-820a364aa737)
![image](https://github.com/user-attachments/assets/eb42279e-d0e7-40f0-8a27-142614838bd8)

youngcw commented on packages/desktop-client/src/hooks/useCategoryScheduleGoalTemplate.ts:
```diff
@@ -0,0 +1,44 @@
+export function useCategoryScheduleGoalTemplate({
+  category,
+}: UseCategoryScheduleGoalTemplateProps): UseCategoryScheduleGoalTemplateResult {
+  const notes = useNotes(category.id);
```
I think it would be better to use the parsed templates that are already stored in the db instead of reparsing the notes.  That way nothing will change when we move away from using the notes for templates.

joel-jeremy replied:
Nice suggestion! Updated to use goal_def instead of notes :)

youngcw commented:
The indicator is more sticky than the templates, so that probably will cause some confusion.

1. Remove the template, rerun templates and the indicator sticks around.  It goes away if the page is refreshed.
2. Disabling templates doesn't remove the indicator

joel-jeremy commented:
> The indicator is more sticky than the templates, so that probably will cause some confusion.
> 
> 1. Remove the template, rerun templates and the indicator sticks around.  It goes away if the page is refreshed.
> 2. Disabling templates doesn't remove the indicator

Should be fixed now :)

youngcw commented:
> > The indicator is more sticky than the templates, so that probably will cause some confusion.
> > 
> > 1. Remove the template, rerun templates and the indicator sticks around.  It goes away if the page is refreshed.
> > 2. Disabling templates doesn't remove the indicator
> 
> Should be fixed now :)

2 got fixed, but 1 is still there.  I guess that one might be hard to fix without constantly checking the goal_def

youngcw commented:
After removing the template and rerunning
![image](https://github.com/user-attachments/assets/e548ebf3-0710-4d22-a374-dcec6fde3638)

joel-jeremy commented:
> After removing the template and rerunning. The balance colors are gone confirming that there is no template. ![image](https://private-user-images.githubusercontent.com/28542559/436237398-e548ebf3-0710-4d22-a374-dcec6fde3638.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNTUyMDAsIm5iZiI6MTc0NTM1NDkwMCwicGF0aCI6Ii8yODU0MjU1OS80MzYyMzczOTgtZTU0OGViZjMtMDcxMC00ZDIyLWEzNzQtZGNlYzZmZGUzNjM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIyVDIwNDgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliZjY1NzM0ODIxMjRkNjk0ZWU3ODc4YTljMGQyNDQxZTcyMTY4OGRkMWFiOTcyNDMwY2NmOGM1MGU2MmZhYzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S7kDICvVv9FnWIc35xv-aOuNEqdxXTJjBRq8HLZyTlQ)
> 
> Im testing on Brave on linux

That's weird, the video I attached shows the color and the indicator disappearing after applying template on the category. Are you using envelope budgeting?

Side: What do you think of putting the indicator on the spent column instead?
![image](https://github.com/user-attachments/assets/ec0c4516-5b1d-49a7-98c9-32bac64cd91c)

youngcw commented:
I was thinking about moving it to the spent column but hadn't mentioned it.  That probably is the best place to put if since I imagine there would be cases where you could see the indicator for multiple months

joel-jeremy commented:
> I was thinking about moving it to the spent column but hadn't mentioned it. That probably is the best place to put it since, I imagine, there would be cases where you could see the indicator for multiple months

Makes sense! I have move it to the spent column

youngcw commented:
The colors are really hard to see on light mode.  And the upcoming purple is hard to see on all themes.

youngcw commented:
Ok, that looks a lot better.  My only concern now is that a spent of 0 next to one of the indicators looks a bit strange where the 0 kinda gets lost and things feel unbalanced.

youngcw commented:
> I tried updating the mobile to have the indicator on spent. Thoughts?
> 
> ![image](https://private-user-images.githubusercontent.com/20313680/436627540-99cf9e08-4ebb-4b19-9806-68a121e74345.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MjQyMzAsIm5iZiI6MTc0NTQyMzkzMCwicGF0aCI6Ii8yMDMxMzY4MC80MzY2Mjc1NDAtOTljZjllMDgtNGViYi00YjE5LTk4MDYtNjhhMTIxZTc0MzQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDE1NTg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmMzE3YzIxNjg1NzRhYTFiZDg1NzMwYWI2MmE5YWEwODlkNjg4MGZmNGY1MjhlZGVjZTRhZjZlZWZjZTQ5YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oFQZ9fj_IxyVrt44CqoCnJc_0lyxcj3kSmjr8yu250M)

Thats probably better

joel-jeremy commented:
> Ok, that looks a lot better. My only concern now is that a spent of 0 next to one of the indicators looks a bit strange where the 0 kinda gets lost and things feel unbalanced.

Personally, I think this is good since it brings attention to the 0 spent like "hey, there's going to be some spending on this soon" but I don't mind changing it if there are better suggestions

joel-jeremy commented:
/update-vrt

joel-jeremy commented:
I realized there can be multiple schedule templates on a single category, so I updated the code to support that:
![image](https://github.com/user-attachments/assets/617c47ed-b139-4753-95a2-138af997aefd)
![image](https://github.com/user-attachments/assets/681a6789-9af3-4989-9527-79dc58092df6)
![image](https://github.com/user-attachments/assets/c3c275df-188e-4061-80bb-92f0e226841b)

I have also added preview transactions for the mobile transactions screen since it's fairly easy to implement. Implementing in desktop is a bit more complex so that can probably be revisited on a separate PR.
![image](https://github.com/user-attachments/assets/5d87f850-4b9e-4231-9136-d85bf163da20)

joel-jeremy commented:
/update-vrt

youngcw commented:
Got this after adding all the default schedules to the same category

Error: Unsupported schedule status for tooltip: scheduled at nQ (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:226:12261) at https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:263:52270 at Array.map (<anonymous>) at https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:263:52194 at kD (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:34255) at HP (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:62173) at Pz (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:61092) at Dz (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:75102) at e9 (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:106615) at Noe (https://deploy-preview-4815.demo.actualbudget.org/static/js/index.KqxHRgjJ.js:57:105673)

joel-jeremy commented:
> Could the indicators be combined instead of having a single indicator for each schedule? Im sure there are people out there that have 5-10+ schedules in a single category.
> 
> Maybe the color could be based on the most urgent of the list missed>due>pending>future(white). And maybe for the icon single>reoccurring. Or just the most urgent schedule is the icon to use
> 
> Im also wondering if it would be better to have the indicator to the right of the spent total. That way the indicators can line up.

Nice idea. I opted to show the missed schedules first, due and then upcoming. When you post a transaction for the missed schedule, the indicator should then update to show the due, and then upcoming until there are no relevant schedules left.

> Im also wondering if it would be better to have the indicator to the right of the spent total. That way the indicators can line up.

I think it's better on the left because if we put it in the right, the category number will no longer align with the group numbers

youngcw commented:
The only thing that may be nice to add is a way to see all the schedules and needed amounts for the month.  There was a feature request for something like that.

Maybe the indicator hover message could list all the upcoming schedules?  But its fine if you don't add that now

joel-jeremy commented:
I've updated the schedule indicators to align at the left most part of the spent column.

> The only thing that may be nice to add is a way to see all the schedules and needed amounts for the month. There was a feature request for something like that.
> 
> Maybe the indicator hover message could list all the upcoming schedules? But its fine if you don't add that now

I think this would be a great addition! A new schedule button could be added on the Month header probably beside the month notes. But that's for another PR.

joel-jeremy commented:
It would also be a nice addition to have a new schedule template which users could run to budget `up to` the total amount of all the schedule templates of the category

i.e.

- Category
  - #template schedule Schedule 1 ($100)
  - #template schedule Schedule 2 ($150)

Applying the above templates should budget to make the category balance $250 (adding on top of any remaining leftover balance). For example, if category already has a $50 leftover balance, applying the template would only budget $200 to meet the $250 schedule amount.

matt-fidd commented on packages/desktop-client/src/hooks/useCategoryScheduleGoalTemplateIndicator.ts:
```diff
@@ -0,0 +1,142 @@
+          locale,
+        });
+      })
+      .reduce((acc, tooltip) => {
+        return `${acc}\n${tooltip}`;
+      }, '');
```
I'm getting an extra blank line at the top of the tooltips because the first iteration of this loop adds '\n' before the first entry.

```suggestion
      .join('\n');
```

matt-fidd reviewed (changes requested):
Very nice! Only one small bug I found to comment on then I'll add my approval.

matt-fidd reviewed (approved):
Nice!
